galaxy/tools/deps/mulled/mulled_update_singularity_containers.py

```diff
@@ -0,0 +1,321 @@
+#!/usr/bin/env python
+
+import requests
+from lxml import html
+import subprocess
+import tarfile
+from ruamel.yaml import YAML
+from io import BytesIO
+#from mulled_build import check_output
+from subprocess import check_output
+import logging
+from shutil import copy
+import pickle
+import json
+import argparse
+from jinja2 import Template
+
+from glob import glob
+
+yaml = YAML()
+yaml.allow_duplicate_keys = True
+
+SINGULARITY_DESTINATION = "/data/0/cvmfs/singularity" # file destination for singularity containers
+SINGULARITY_INSTALL = "/opt/singularity/bin/singularity" # location at which singularity is installed, could be something else like /usr/local/bin/singularity
+QUAY_API_ENDPOINT = 'https://quay.io/api/v1/repository'
+
+def get_quay_containers():
+    """
+    Gets all quay containers in the biocontainers repo
+    # >>> lst = get_quay_containers()
+    # >>> 'samtools:latest' in lst
+    # True
+    # >>> 'abricate:0.4--pl5.22.0_0' in lst
+    # True
+    # >>> 'samtools' in lst
+    # False
+    """
+    containers = []
+
+    repos_parameters = {'public': 'true', 'namespace': 'biocontainers'}
+    repos_headers ={'Accept-encoding': 'gzip', 'Accept': 'application/json'}
+    repos_response = requests.get(QUAY_API_ENDPOINT, headers=repos_headers, params=repos_parameters, timeout=12)
+
+    repos = repos_response.json()['repositories']
+    #repos = [n['name'] for n in repos]
+
+    for repo in repos:
+        logging.info(repo)
+        tags_response = requests.get("%s/biocontainers/%s" % (QUAY_API_ENDPOINT, repo['name']))
+        tags = tags_response.json()['tags']
+        for tag in tags:
+            containers.append('%s:%s' % (repo['name'], tag))
+
+    return containers
+
+def get_singularity_containers():
+    """
+    Gets all existing singularity containers from "https://depot.galaxyproject.org/singularity/"
+    # >>> lst = get_singularity_containers()
+    # >>> 'aragorn:1.2.36--1' in lst
+    # True
+    # >>> 'znc:latest' in lst
+    # False
+
+    """
+    index_url = "https://depot.galaxyproject.org/singularity/"
+    index = requests.get(index_url)
+    #l = response.text.split('\n')
+    tree = html.fromstring(index.content)
+    containers = tree.xpath('//a/@href')
+    containers = [container.replace('%3A', ':') for container in containers]
+    return containers
+
+def get_missing_containers(quay_list, singularity_list, blacklist_file=None):
+    """
+    Returns list of quay containers that do not exist as singularity containers. Files stored in a blacklist will be ignored
+    # >>> lst = get_missing_containers()
```

Comment:
get missing containers should get here two lists and then you can include the tests.